runtime/src/accounts_db.rs

```diff
@@ -2199,6 +2234,44 @@ pub mod tests {
         assert_eq!(db0.load_slow(&ancestors, &key), Some((account0, 0)));
     }
 
+    #[test]
+    fn test_remove_unrooted_slot() {
+        let unrooted_slot = 9;
+        let db = AccountsDB::new(Vec::new());
+        let key = Pubkey::default();
+        let account0 = Account::new(1, 0, &key);
+        let ancestors: HashMap<_, _> = vec![(unrooted_slot, 1)].into_iter().collect();
+        db.store(unrooted_slot, &[(&key, &account0)]);
+        db.bank_hashes
+            .write()
+            .unwrap()
+            .insert(unrooted_slot, BankHashInfo::default());
+        assert!(db
+            .accounts_index
+            .read()
+            .unwrap()
+            .get(&key, &ancestors)
+            .is_some());
+        assert!(db.load_slow(&ancestors, &key).is_some());
+
+        // Purge the slot
+        db.remove_unrooted_slot(unrooted_slot);
+        assert!(db.load_slow(&ancestors, &key).is_none());
+        assert!(db.bank_hashes.read().unwrap().get(&unrooted_slot).is_none());
+        assert!(db.storage.read().unwrap().0.get(&unrooted_slot).is_none());
+        assert!(db
+            .accounts_index
+            .read()
+            .unwrap()
+            .get(&key, &ancestors)
+            .is_none());
+
+        // Test we can store for the same slot again and get the right information
+        let account0 = Account::new(2, 0, &key);
+        db.store(unrooted_slot, &[(&key, &account0)]);
+        assert_eq!(db.load_slow(&ancestors, &key).unwrap().0.lamports, 2);
+    }
```

Comment:
How about a test that does this operation and then a snapshot save/restore?

Reply:
added one here leveraging some nice existing test functions: https://github.com/solana-labs/solana/pull/9805/files#diff-2099c5256db4eb5975c8834af38f6456R2293

How does that look?